### PR TITLE
Refactored attribute value accessors

### PR DESCRIPTION
**Pull request by LesTR**
Introduce attribute value accessors which allows manipulation with attribute values.

### CODE REVIEW

File: core/src/main/java/cz/o2/proxima/scheme/ValueSerializer.java [hunk 1]
```diff
@@ -110,4 +110,14 @@ default String getLogString(T value) {
...
   }
+
+  /**
+   * Provide {@link AttributeValueAccessor} for give attribute value.
+   *
+   * @return value accessor
+   */
+  default <V> AttributeValueAccessor<T, V> getValueAccessor() {
+    throw new UnsupportedOperationException(
+        getClass() + " is not ported to provide a ValueAccessor. " + "Please fill issue.");
```
Comment (je-ik) [inline core/src/main/java/cz/o2/proxima/scheme/ValueSerializer.java:121]: nit: Probably no need to separate the two strings. But can be fixed by different PR (I think I already did that somewhere).

**Review by je-ik — APPROVED** [review]
LGTM!

File: core/src/main/java/cz/o2/proxima/scheme/AttributeValueAccessor.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+import java.io.Serializable;
+
+/**
+ * Interface for value accessors allowed create and get value of attribute
+ *
+ * @param <T> input type
+ * @param <V> output type
+ */
+@Experimental
+public interface AttributeValueAccessor<T, V> extends Serializable {
```
Comment (dmvk) [inline core/src/main/java/cz/o2/proxima/scheme/AttributeValueAccessor.java:28]:
```suggestion
public interface AttributeValueAccessor<InputT, OutputT> extends Serializable {
```

File: core/src/main/java/cz/o2/proxima/scheme/AttributeValueAccessors.java [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+    public T createFrom(V object) {
+      return createFromConvertCallback.apply(object);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      return o != null && getClass() == o.getClass();
```
Comment (dmvk) [inline core/src/main/java/cz/o2/proxima/scheme/AttributeValueAccessors.java:283]: What is the reasoning for this equals implementation?

If we don't need this, I'd rather see it removed.

File: core/src/main/java/cz/o2/proxima/scheme/AttributeValueAccessors.java [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
```
Comment (dmvk) [inline core/src/main/java/cz/o2/proxima/scheme/AttributeValueAccessors.java:313]: Same as above

File: core/src/main/java/cz/o2/proxima/scheme/ValueSerializer.java (hunk 1 above)
Comment (dmvk) [reply core/src/main/java/cz/o2/proxima/scheme/ValueSerializer.java:121]: nit: same applies to schema descriptor above

File: scheme/proto/src/main/java/cz/o2/proxima/scheme/proto/ProtoMessageValueAccessor.java [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+  @SuppressWarnings("unchecked")
+  private Object mapValue(FieldDescriptor descriptor, Object value) {
+    if (descriptor.getJavaType().equals(JavaType.MESSAGE)) {
+      return StructureValue.of((Map<String, Object>) value);
+    }
+    return value;
+  }
+
+  @SuppressWarnings("unchecked")
+  private <X, Y> AttributeValueAccessor<X, Y> createFieldAccessor(
```
Comment (dmvk) [inline scheme/proto/src/main/java/cz/o2/proxima/scheme/proto/ProtoMessageValueAccessor.java:149]:
```suggestion
  private <InputT, OutputT> AttributeValueAccessor<InputT, OutputT> createFieldAccessor(
```

This should definitely change to make thinks more readable.

File: scheme/proto/src/main/java/cz/o2/proxima/scheme/proto/ProtoSerializerFactory.java [hunk 6]
```diff
@@ -84,6 +85,8 @@ public String getAcceptableScheme() {
 
     @Nullable private transient SchemaTypeDescriptor<M> valueSchemaDescriptor;
 
+    @Nullable private ProtoMessageValueAccessor<M> accessor;
```
Comment (dmvk) [inline scheme/proto/src/main/java/cz/o2/proxima/scheme/proto/ProtoSerializerFactory.java:88]: transient? Also, are there any assumptions / concerns about thread safety?

File: scheme/proto/src/main/java/cz/o2/proxima/scheme/proto/ProtoSerializerFactory.java [hunk 7]
```diff
@@ -158,5 +161,14 @@ public M fromJsonValue(String json) {
       }
       return valueSchemaDescriptor;
     }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <V> AttributeValueAccessor<M, V> getValueAccessor() {
```
Comment (dmvk) [inline scheme/proto/src/main/java/cz/o2/proxima/scheme/proto/ProtoSerializerFactory.java:167]:
```suggestion
    public <OutputT> AttributeValueAccessor<MessageT, OutputT> getValueAccessor() {
```

File: scheme/proto/src/main/java/cz/o2/proxima/scheme/proto/ProtoMessageValueAccessor.java [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+  private Object prepareFieldValue(FieldDescriptor descriptor, Object object) {
+    if (descriptor.isRepeated()) {
+      if (List.class.isAssignableFrom(object.getClass())) {
+        return ((List<Object>) object)
+            .stream()
+            .map(v -> mapValue(descriptor, v))
+            .collect(Collectors.toList());
+      } else {
+        // Create new list in case when originally scala field changed to repeated
+        return Collections.singletonList(mapValue(descriptor, object));
```
Comment (dmvk) [inline scheme/proto/src/main/java/cz/o2/proxima/scheme/proto/ProtoMessageValueAccessor.java:134]: this looks fragile, can you please elaborate little bit about this edge case?

**Review by dmvk — APPROVED** [review]
This is a really nicely scoped, well prepared PR 👏

I have few (mostly cosmetic concerns), I've outlined in the comments. I don't see any major issues here (maybe `equals` implementation in `ValueAccessors` could bite us in the future).

LGTM overall 👍

File: core/src/main/java/cz/o2/proxima/scheme/ValueSerializer.java (hunk 1 above)
Comment (LesTR) [reply core/src/main/java/cz/o2/proxima/scheme/ValueSerializer.java:121]: applied on all.

File: scheme/proto/src/main/java/cz/o2/proxima/scheme/proto/ProtoMessageValueAccessor.java (hunk 8 above)
Comment (LesTR) [reply scheme/proto/src/main/java/cz/o2/proxima/scheme/proto/ProtoMessageValueAccessor.java:134]: There is a typo in word "scala", it should be scalar (fixed). Basically protobuf allows you to change field name from for example `string field` to `repeated string field` (same for opposite direction) but you will have value stored in original form. In than case we need to create an list from scalar value (to be able to call addRepeated method on builder). For details please see `testCreateProtoWhereOptionalFieldChangedToRepeated` and `testCreateProtoWhereRepeatedFieldChangedToOptional` in `ProtoMessageValueAccessorTest`

Comment (dmvk) [reply scheme/proto/src/main/java/cz/o2/proxima/scheme/proto/ProtoMessageValueAccessor.java:134]: 👍 Ok, now it makes more sense :) Maybe we could change `List.isAssignable..` to `Collection.isAssignable...` to relax the condition?

File: core/src/main/java/cz/o2/proxima/scheme/AttributeValueAccessors.java (hunk 3 above)
Comment (LesTR) [reply core/src/main/java/cz/o2/proxima/scheme/AttributeValueAccessors.java]: removed.

**Pull request merged**